blog/2024-03-08-structuring-a-large-app.md

```diff
@@ -0,0 +1,391 @@
+
+# State Management
+
+As of Reflex 0.4.3, any event handler can get access to an instance of any other
+substate via the `get_state` API. From a practical perspective, this means that
+state can be split up into smaller pieces without requiring a complex
+inheritance hierarchy to share access to other states.
+
+In previous releases, if an app wanted to store settings in `SettingsState` with
+a page or component for modifying them, any other state with an event handler
+that needed to access those settings would have to inherit from `SettingsState`,
+even if the other state was mostly orthogonal. The other state would also now
+always have to load the settings, even for event handlers that didn't need to
+access them.
+
+A better strategy is to load the desired state on demand from only the event
+handler which needs access to the substate.
+
+## example_big_app/components/settings.py
+
+```python
+import reflex as rx
+
+
+class SettingsState(rx.State):
+    refresh_interval: int = 15
+    auto_update: bool = True
+    prefer_plain_text: bool = True
+    posts_per_page: int = 20
+
+
+def settings_dialog():
+    return rx.dialog(...)
+```
+
+## example_big_app/pages/posts.py
+
+```python
+import reflex as rx
+
+from ..models import Post
+from ..template import template
+from ..components.settings import SettingsState
+
+
+class PostsState(rx.State):
+    refresh_tick: int
+    page: int
+    posts: list[Post]
+
+    async def on_load(self):
+        settings = await self.get_state(SettingsState)
+        if settings.auto_update:
+            self.refresh_tick = settings.refresh_interval * 1000
+        else:
+            self.refresh_tick = 0
+
+    async def tick(self, _):
+        settings = await self.get_state(SettingsState)
+        with rx.session() as session:
+            q = Post.select.offset(self.page * settings.posts_per_page).limit(settings.posts_per_page)
+            self.posts = q.all()
+
+    def go_to_previous(self):
+        if self.page > 0:
+            self.page = self.page - 1
+
+    def go_to_next(self):
+        if self.posts:
+            self.page = self.page + 1
+
+
+@rx.page(route="/posts", on_load=PostsState.on_load)
+@template
+def posts():
+    return rx.vstack(
+        rx.foreach(PostsState.posts, post_view),
+        rx.hstack(
+            rx.button("< Prev", on_click=PostsState.go_to_previous),
+            rx.button("Next >", on_click=PostsState.go_to_next),
+            justify="between",
+        ),
+        rx.moment(interval=PostsState.refresh_tick, on_change=PostsState.tick, display="none"),
+        width="100%",
+    )
+```
+
+# Component Reusability
+
+The primary mechanism for reusing components in Reflex is to define a function that returns
+the component, then simply call it where that functionality is needed.
+
+Component functions typically should not take any State classes as arguments, but prefer
+to import the needed state and access the vars on the class directly.
+
+## functools.lru_cache
```

Comment:
I think we can have this in the docs, but not needed as part of "project structure"

Reply:
removed from post